### PR TITLE
fix: `trace_debugTransaction` is inconsistent with geth's responses for tracer 'callTracer'

### PR DESCRIPTION
**Pull request by qiweiii**
## Motivation

fix: #6882

## Solution

Now `geth_trace` returns DefaultFrame:

https://github.com/foundry-rs/foundry/blob/c6790a02bb80b163469e2e3bb0b16ae95ad7334f/crates/anvil/src/eth/backend/mem/storage.rs#L419-L420

Changed to return GethTracer and check if tracer is CallTracer type

### CODE REVIEW

File: crates/anvil/src/eth/backend/mem/storage.rs [hunk 1]
```diff
@@ -416,13 +416,33 @@ impl MinedTransaction {
...
-            .geth_traces(
-                self.receipt.gas_used().as_u64(),
-                self.info.out.clone().unwrap_or_default().0.into(),
-                opts,
+    pub fn geth_trace(&self, opts: GethDebugTracingOptions) -> GethTrace {
+        let GethDebugTracingOptions { config, tracer, tracer_config, .. } = opts;
+
+        if let Some(GethDebugTracerType::BuiltInTracer(GethDebugBuiltInTracerType::CallTracer)) =
+            tracer
+        {
```
Comment (mattsse) [inline crates/anvil/src/eth/backend/mem/storage.rs:424]: let's do a match here because eventually, we want to add the remaining tracers as well

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/anvil/src/eth/backend/mem/storage.rs (hunk 1 above)
Comment (qiweiii) [reply crates/anvil/src/eth/backend/mem/storage.rs]: hi @mattsse, I am not sure how to handle `PreStateTracer` and `JsTracer` case, seems we need to change `geth_trace` return to an Result type and also add an `inspect` method.

Some pointers would be appreciated! Also I feel a bit overwhelmed to impl this, if you have time, feel free to continue in this branch

File: crates/anvil/src/eth/backend/mem/storage.rs [hunk 2]
```diff
@@ -416,21 +425,46 @@ impl MinedTransaction {
...
+                GethDebugTracerType::JsTracer(code) => {
+                    let config = tracer_config.into_json();
+
+                    let mut inspector = JsInspector::new(code, config)?;
+                    let (res, env, db) = inspect_and_return_db(db, env, &mut inspector)?;
+
+                    let state = res.state.clone();
+                    let result = inspector.json_result(res, &env, db)?;
+                    Ok((GethTrace::JS(result), state))
+                }
```
Comment (mattsse) [inline crates/anvil/src/eth/backend/mem/storage.rs:465]: this we can't support yet, so this should not do anything

File: crates/anvil/src/eth/backend/mem/storage.rs [hunk 3]
```diff
@@ -416,21 +425,46 @@ impl MinedTransaction {
         })
     }
 
-    pub fn geth_trace(&self, opts: GethDebugTracingOptions) -> GethTrace {
+    pub fn geth_trace<DB: DatabaseRef>(
+        &self,
+        opts: GethDebugTracingOptions,
+        env: Env,
+        db: DB,
```
Comment (mattsse) [inline crates/anvil/src/eth/backend/mem/storage.rs:432]: does we don't need, or rather can't use yet

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/anvil/src/eth/backend/mem/storage.rs (hunk 1 above)
Comment (mattsse) [reply crates/anvil/src/eth/backend/mem/storage.rs]: let's ignore them for now, we'd need to figure out a way to replay transaction first, so let's just leave these match arms empty/return an error

**Comment by prasincs** [comment]
tried to build this and ran into same issues as clippy is reporting. Looks like integration tests are failing too

**Comment by qiweiii** [comment]
@prasincs I just fixed the code, it would be great if you could help test the calltracer

**Comment by prasincs** [comment]
@qiweiii I ran a quick test and it seems to preserve behavior, I only checked gas for an erc20 transfer because rest of fields seem to decode correctly before https://github.com/prasincs/anvil-tests/blob/main/README.md

**Comment by qiweiii** [comment]
Maybe it's related to code in `geth_call_traces` which is in this [repo](https://github.com/paradigmxyz/evm-inspectors/blob/5ee90076f88e64a059e4c0c268929cf9ec91ed6d/src/tracing/builder/geth.rs#L117)

**Comment by prasincs** [comment]
@qiweiii what's a good way to unblock this? this seems to be okay for native eth transfers and some basic cases but I can do some deeper tests with anvil if required?

**Comment by chase-guardrail** [comment]
I'm interested in this as well

File: crates/anvil/src/eth/backend/mem/storage.rs [hunk 4]
```diff
@@ -415,13 +418,44 @@ impl MinedTransaction {
...
+                        return FourByteFrame::from(inspector).into()
+                    }
+                    GethDebugBuiltInTracerType::CallTracer => {
+                        let call_config = tracer_config.into_call_config().unwrap();
+
+                        return GethTraceBuilder::new(
+                            self.info.traces.clone(),
+                            TracingInspectorConfig::from_geth_config(&config),
+                        )
+                        .geth_call_traces(call_config, self.receipt.gas_used())
```
Comment (chase-guardrail) [inline crates/anvil/src/eth/backend/mem/storage.rs:438]: I think this needs to be converted:

.geth_call_traces(call_config, self.receipt.gas_used().to::<u64>())

**Comment by qiweiii** [comment]
@prasincs Does this work now? I have trouble with setting it up and test it locally 🤦🏻

**Comment by chase-guardrail** [comment]
> @prasincs Does this work now? I have trouble with setting it up and test it locally 🤦🏻

Been working for me locally 👍

File: crates/anvil/src/eth/backend/mem/storage.rs [hunk 5]
```diff
@@ -415,13 +418,44 @@ impl MinedTransaction {
...
+
+        if let Some(tracer) = tracer {
+            return match tracer {
+                GethDebugTracerType::BuiltInTracer(tracer) => match tracer {
+                    GethDebugBuiltInTracerType::FourByteTracer => {
+                        let inspector = FourByteInspector::default();
+                        return FourByteFrame::from(inspector).into()
+                    }
+                    GethDebugBuiltInTracerType::CallTracer => {
+                        let call_config = tracer_config.into_call_config().unwrap();
```
Comment (mattsse) [inline crates/anvil/src/eth/backend/mem/storage.rs:432]: this can fail if argument is invalid

**Comment by prasincs** [comment]
> @prasincs Does this work now? I have trouble with setting it up and test it locally 🤦🏻

I couldn't build the container for full e2e test but it works for some adhoc tests I've done locally and from tests in PR. Once it merges I can validate with the container

**Review by onbjerg — COMMENTED** [review]
need to pass clippy, otherwise lgtm

**Comment by prasincs** [comment]
Can someone look at the MSVC integration test issues, it looks like some specific remapping issue that I'm not familiar with

**Comment by crystalbit** [comment]
Hey guys! We have built anvil from this PR (commit `7250765a9dd8737f36c1ba90a06e337a9ba7d29a`) and seems there is no callTracer in usual way

90% that I'm wrong somewhere, could you pls kindly check my screenshot?

I run anvil with `--fork-url https://ipc.kras.fluence.dev`

image

**Comment by qiweiii** [comment]
@mattsse can let someone pick up this PR, I'm afraid I may not have time to finish this up anytime soon

**Comment by qiweiii** [comment]
hi @crystalbit, seems `debug_tracerCall` has problem, but if you use `debug_traceTransaction`, which this PR fix, it will work:

image

maybe fix `debug_tracerCall` in another PR? @mattsse 

also can we get this merged soon?

**Comment by georgeroman** [comment]
Hey everyone, any updates on this?

**Comment by d3or** [comment]
Hey everyone, are the MSVC integration test issues still a problem here? I see that all the checks have passed

**Review by DaniPopes — APPROVED** [review]
Thanks! Sorry for the delay.

**Pull request merged**